kwasm-lemmas.md

```diff
@@ -308,68 +310,51 @@ We want to make the variant explicit, so we introduce the following helper, whic
       requires notBool isInt(I)
 ```
 
-We can now state the intended semantics of `#get`.
-
-```k
-    rule LOW            <=Int #get(BM, ADDR) => true
-      requires LOW <=Int 0
-       andBool #isByteMap(BM) [simplification]
-
-    rule #get(BM, ADDR)  <Int HIGH           => true
-      requires 256 <=Int HIGH
-       andBool #isByteMap(BM) [simplification]
-```
-
 With this invariant encoded, we can introduce the following lemma.
 
 ```k
     rule #isByteMap(BMAP) impliesBool (0 <=Int #get(BMAP, IDX) andBool #get(BMAP, IDX) <Int 256) => true [smt-lemma]
 ```
 
-From the semantics, it should be clear that setting the index in a bytemap to the value already contained there will leave the map unchanged.
-Conversely, setting an index in a map to a value `VAL` and then retrieving the value at that index will yield `VAL`.
-A getting operation can safeley ignore setting operation on unrelated indices.
-
-```k
-    rule #set(BMAP, IDX, #get(BMAP, IDX)) => BMAP [smt-lemma]
-
-    rule #get(#set(BMAP, IDX', VAL), IDX) => VAL             requires         IDX ==Int IDX'
-    rule #get(#set(BMAP, IDX', VAL), IDX) => #get(BMAP, IDX) requires notBool IDX ==Int IDX'
-```
-
 TODO: We should inspect the two functions `#getRange` and `#setRange` closer.
 They are non-trivial in their implementation, but the following should obviously hold from the intended semantics.
 
 ```k
-    rule #setRange(BM, EA, #getRange(BM, EA, WIDTH), WIDTH) => BM
-    rule #setRange(BM, EA, VAL, WIDTH) => #set(BM, EA, #wrap(1, VAL))
-      requires WIDTH ==Int 1
-      [simplification]
-
-    rule #getRange(BM, ADDR, WIDTH) modInt BYTE_SIZE => #get(BM, ADDR)
-      requires BYTE_SIZE ==Int 256
-       andBool notBool (WIDTH <=Int 0)
+    rule #getRange(BM, ADDR, WIDTH) >>Int SHIFT => #getRange(BM, ADDR +Int 1, WIDTH -Int 1) >>Int (SHIFT -Int 8)
+      requires 0 <=Int ADDR
+       andBool 0  <Int WIDTH
+       andBool 8 <=Int SHIFT
        andBool #isByteMap(BM)
       [simplification]
 
-    rule #wrap(N, #getRange(BM, ADDR, WIDTH)) => #get(BM, ADDR)
-      requires N ==Int 1
-       andBool notBool (WIDTH <=Int 0)
+    rule #getRange(BM, ADDR, WIDTH) modInt MAX => #getRange(BM, ADDR, WIDTH)
+      requires 0 <Int MAX andBool 0 <Int WIDTH
+       andBool 2 ^Int (8 *Int WIDTH) <=Int MAX
        andBool #isByteMap(BM)
       [simplification]
 
-    rule #wrap(N, #getRange(BM, ADDR, WIDTH)) => #getRange(BM, ADDR, N)
-      requires 0 <=Int N andBool N <=Int WIDTH
+    rule #getRange(BM, ADDR, WIDTH) modInt MAX => #getRange(BM, ADDR, WIDTH -Int 1) modInt MAX
+      requires 0 <Int MAX andBool 0 <Int WIDTH
+       andBool MAX <=Int 2 ^Int (8 *Int (WIDTH -Int 1))
```

Comment:
This is only sound _in general_ when `MAX` is a power of 2. (But it should be sound for any case that can show up in our semantics.) Counterexample to the above:
```
BM := (0 |-> 0) (1 |-> 1)
ADDR := 0
WIDTH := 2
==> #getRange(BM, ADDR, WIDTH) == 256
and #getRange(BM, ADDR, WIDTH -Int 1) == 0

MAX = 3
==> MAX <=Int 2 ^Int (8 * Int (WIDTH -Int1)) == 2^8
==> LHS == 1 and RHS == 0

Reply:
Hmmm, good catch, I'll look into fixing it.